### ISSUE TITLE
AIX OOMs in extended.system SharedClasses MultiThread testing

### ISSUE THREAD

**Issue by pshipton** [eclipse/openj9#8842]
"java/lang/OutOfMemoryError" "Failed to create a thread: retVal -1073741830, errno 11"
11 EAGAIN The limit on the number of threads per process has been reached.
-1073741830 == 0xBFFFFFFA, negated is 0x40000006
J9THREAD_ERR_THREAD_CREATE_FAILED 6
J9THREAD_ERR_OS_ERRNO_SET 0x40000000

https://ci.eclipse.org/openj9/job/Test_openjdk11_j9_extended.system_ppc64_aix_Nightly/311/
aix71-p8-4
SharedClasses.SCM01.MultiThread_0 
SharedClasses.SCM23.MultiThread_0

https://ci.eclipse.org/openj9/job/Test_openjdk14_j9_extended.system_ppc64_aix_Nightly/14/
aix71-p8-1
SharedClasses.SCM01.MultiThread_0
SharedClasses.SCM23.MultiThread_0
SharedClasses.SCM23.MultiThreadMultiCL_0

https://ci.eclipse.org/openj9/job/Test_openjdk8_j9_extended.system_ppc64_aix_Nightly/313/
SharedClasses.SCM01.MultiThread_0
SharedClasses.SCM23.MultiThread_0

Changes from previous passing build (from jdk11).

https://github.com/eclipse/openj9/compare/a1ed808...c209fa5
https://github.com/eclipse/openj9-omr/compare/b03105e...79f6485
https://github.com/ibmruntimes/openj9-openjdk-jdk11/compare/b1d6957...7d1badb

https://github.com/AdoptOpenJDK/openjdk-tests/compare/8737892...e459da2

**Comment by pshipton** [comment]
@jdekonin could the machines have changed, this doesn't seem to have been caused by a JVM or test change.

**Comment by pshipton** [comment]
Ah, nm, I missed a change, this is caused by https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/272
@zl-wang @shanchao95 
Zulian, this seems problematic. Is it expected that "MALLOCOPTIONS=multiheap,buckets" would reduce the number of threads that can be created?

**Comment by zl-wang** [comment]
no, not aware of that.  especially, it doesn't make sense for 64bit.  let me post a question to AIX folks.

**Comment by zl-wang** [comment]
i assumed errno==4 being returned upon pthread_create call?   according to the message above

**Comment by pshipton** [comment]
Why 4 (INTR)? pthread_create is returning 11 (EAGAIN), which is "The limit on the number of threads per process has been reached.".

We did see a similar problem creating threads because of the data limit, but we modified the VM to get the data soft limit to the hard limit.
https://github.com/eclipse/openj9/pull/8367

**Comment by zl-wang** [comment]
i was looking at  "J9THREAD_ERR_OS_ERRNO_SET 0x40000000"   ...  implying 4 seemed.

**Comment by zl-wang** [comment]
cryptic error message

**Comment by zl-wang** [comment]
AIX kernel folks responded:
pthread_create returns EAGAIN for exceeding the thread count ulimit or failure to allocate (malloc) memory for internal pthread data structures.

I would say:
Let's dig a little deeper into the javacore file:  what is the ulimit? how many threads are currently active?

short of ulimit issue, it is unimaginable that malloc failed for 64bit.

if we can limit good & bad drivers to be without/with that PR, then we can experiment further with settings:
MALLOCOPTIONS=multiheap:4
MALLOCOPTIONS=multiheap:8
MALLOCOPTIONS=multiheap:16
(dropping ,buckets  which I didn't see in the manual anyway)

**Comment by zl-wang** [comment]
by the way, default is 32 heaps (max as well) without :n.

**Comment by zl-wang** [comment]
I do see buckets option later on. Let's play with/without it as well ... for example:
MALLOCOPTIONS=multiheap:4,buckets

**Comment by pshipton** [comment]
https://ci.eclipse.org/openj9/view/Test/job/Grinder/666 - multiheap,buckets - failed

https://ci.eclipse.org/openj9/view/Test/job/Grinder/661 - multiheap:4 - passed
https://ci.eclipse.org/openj9/view/Test/job/Grinder/662 - multiheap:4,buckets - passed
https://ci.eclipse.org/openj9/view/Test/job/Grinder/663 - multiheap:8 - passed
https://ci.eclipse.org/openj9/view/Test/job/Grinder/664 - multiheap:8,buckets - passed
https://ci.eclipse.org/openj9/view/Test/job/Grinder/667 - multiheap:16 - 
https://ci.eclipse.org/openj9/view/Test/job/Grinder/668 - multiheap:16, buckets -

**Comment by zl-wang** [comment]
16 and 16,buckets passed too.  could you launch explicit 32 and 32,buckets build/tests?

**Comment by zl-wang** [comment]
for pinning down the symptom, maybe it is good to launch 24,buckets ...  28,buckets, and 29/30/31,buckets.  Let's see when it cracks.

**Comment by pshipton** [comment]
https://ci.eclipse.org/openj9/view/Test/job/Grinder/671 - multiheap:32 - passed
https://ci.eclipse.org/openj9/view/Test/job/Grinder/672 - multiheap:32,buckets - failed
https://ci.eclipse.org/openj9/view/Test/job/Grinder/673 - multiheap:28,buckets - passed
https://ci.eclipse.org/openj9/view/Test/job/Grinder/674 - multiheap:24,buckets - failed

I'm only running one iteration of each test. Now 24,buckets has failed, but 28,buckets passed.

**Comment by zl-wang** [comment]
that seemed indicating EAGAIN can happen unpredictably, depending on timing, for pthread_create call.  assuming the same number of threads created in each run, that leaves only one possibility that virtual or data memory ulimit is exceeded unpredictably with massive concurrent malloc/free going on.  it could be momentarily exceeded though.   

32,buckets failed around 14700 threads; while 24,buckets failed around 18000 threads. 32,buckets have more fragmentations leading to the momentarily exceeding.  it makes sense to me.  ",buckets" option makes it doubly fragmented with many concurrent malloc/free.

could you find out the ulimit for virtual memory and data memory for these runs? once we know the reason, we can decide on a fix.

**Comment by zl-wang** [comment]
https://ci.eclipse.org/openj9/view/Test/job/Grinder/666  (effectively  32,buckets)  failed around 17400 threads, while the later 32,buckets run failed around 14700 threads.  indeed, it is unpredictable.  my theory seemed standing ...

**Comment by pshipton** [comment]
https://ci.eclipse.org/openj9/view/Test/job/Grinder/676 - multiheap:16,buckets x 3 - passed
https://ci.eclipse.org/openj9/view/Test/job/Grinder/677 - multiheap:20,buckets x 3 - passed
https://ci.eclipse.org/openj9/view/Test/job/Grinder/678 - multiheap:22,buckets x 3 - passed

**Comment by zl-wang** [comment]
if all are unlimited, we might need to bring it forward to AIX kernel team.  Logically at least.  Depending on how to interpret EAGAIN.

**Comment by pshipton** [comment]
From https://ci.eclipse.org/openj9/view/Test/job/Grinder/674 - multiheap:24,buckets javacore
```
1CIUSERLIMITS  User Limits (in bytes except for NOFILE and NPROC)
NULL           ------------------------------------------------------------------------
NULL           type                            soft limit           hard limit
2CIUSERLIMIT   RLIMIT_AS                        unlimited            unlimited
2CIUSERLIMIT   RLIMIT_CORE                      unlimited            unlimited
2CIUSERLIMIT   RLIMIT_CPU                       unlimited            unlimited
2CIUSERLIMIT   RLIMIT_DATA                      unlimited            unlimited
2CIUSERLIMIT   RLIMIT_FSIZE                     unlimited            unlimited
2CIUSERLIMIT   RLIMIT_NOFILE                    unlimited            unlimited
2CIUSERLIMIT   RLIMIT_THREADS                   unlimited            unlimited
2CIUSERLIMIT   RLIMIT_NPROC                     unlimited            unlimited
2CIUSERLIMIT   RLIMIT_RSS                        33554432            unlimited
2CIUSERLIMIT   RLIMIT_STACK                    4294967296           4294967296
```

**Comment by zl-wang** [comment]
it is likely RSS being exceeded momentarily.   didn't see "virtual memory"  (-v)  though.

**Comment by zl-wang** [comment]
make RSS unlimited and retry ...

the eventual fix can go:  without  ,buckets.  need to confirm with performance benefit/data of ",buckets" from security test cases.

**Comment by pshipton** [comment]
> make RSS unlimited and retry ...

Is there a particular setting I should try with RSS unlimited?

> the eventual fix can go: without ,buckets

So I'll try a run with `MALLOCOPTIONS=multiheap` then?

**Comment by zl-wang** [comment]
ulimit -m unlimited  (for RSS unlimited)

yes, we can try a few multiheap alone runs.

**Comment by pshipton** [comment]
> ulimit -m unlimited (for RSS unlimited)

yes, but with which testcase, `multiheap:24,buckets`?

**Comment by zl-wang** [comment]
> yes, but with which testcase, multiheap:24,buckets?

yes, with any of the previously failing setting(s), e.g.  multiheap,buckets

**Comment by pshipton** [comment]
With ulimit -m unlimited:

https://ci.eclipse.org/openj9/view/Test/job/Grinder/679 - multiheap:24,buckets x 3
https://ci.eclipse.org/openj9/view/Test/job/Grinder/680 - multiheap:32,buckets
https://ci.eclipse.org/openj9/view/Test/job/Grinder/681 - multiheap,buckets

Without changing ulimit:
https://ci.eclipse.org/openj9/view/Test/job/Grinder/682 - multiheap x 3

**Comment by pshipton** [comment]
Updated results in the previous comment.

**Comment by pshipton** [comment]
@zl-wang is someone looking at the performance of using `MALLOCOPTIONS=multiheap` ?

**Comment by zl-wang** [comment]
@chao.shan@ibm.com  could you compare on AIX crypto performance test cases between MALLOCOPTIONS=multiheap vs. MALLOCOPTIONS=multiheap,buckets?

**Comment by shanchao95** [comment]
@pshipton @zl-wang Sure, here is AIX crypto performance and compare for `NO MALLOCOPTION`, `multiheap `, `multiheap, buckets` .  )
[performance.xlsx](https://github.com/eclipse/openj9/files/4349085/performance.xlsx)

**Comment by zl-wang** [comment]
Thanks for the performance data.  Summary:
1) MALLOCOPTIONS setting(s) almost have no performance degradation over no setting, except one case with CBC-decrypt-stream-nopadding 16-thread;
2) multiheap setting got most of the performance improvement there is;
3) multiheap,buckets might have less than 10% performance advantage over multiheap setting on average ... there are a few cases of regression too (up to -25%).

Plus, multiheap,buckets used more memory and failed this PR's test case.

I would settle for multiheap setting as a fix.  @shanchao95  please fix it as indicated.

**Comment by pshipton** [comment]
Fixing in
https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/389
https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/390
https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/279
https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/280
https://github.com/ibmruntimes/openj9-openjdk-jdk14/pull/29
https://github.com/ibmruntimes/openj9-openjdk-jdk14/pull/30
https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/181

**Comment by keithc-ca** [comment]
All the pull requests mention in https://github.com/eclipse/openj9/issues/8842#issuecomment-601313084 are now merged. Is it time to close this?

**Comment by pshipton** [comment]
I did all my testing using jdk8, assuming the problem would be the same across versions. It seems the problem on jdk8 is resolved by the change, but not jdk11 or jdk14. I'll have to run more testing.

**Comment by pshipton** [comment]
jdk11

https://ci.eclipse.org/openj9/view/Test/job/Grinder/698 - multiheap:4
https://ci.eclipse.org/openj9/view/Test/job/Grinder/699 - multiheap:8

**Comment by zl-wang** [comment]
all passed.   please try multiheap:24/28/30 multiple times.  i assumed we don't need to go through another round of performance test, since the heap number is definitely more than 16 (which is the max thread number of existing test cases).

@shanchao95  maybe you need to change your test cases to increase the thread number to 24 and 32, and test against multiheap:24/28/30/32 to see whether there are significant differences.

**Comment by pshipton** [comment]
@zl-wang would one of the above be acceptable? Let me know what we should be aiming for and I can run more tests.

**Comment by pshipton** [comment]
Sorry, didn't see your comment before I posted mine.

**Comment by pshipton** [comment]
Given previous results, I suspect the following will work

https://ci.eclipse.org/openj9/view/Test/job/Grinder/701 - multiheap:32 x 3

**Comment by zl-wang** [comment]
unlikely to succeed.  multiheap is multiheap:32  (the default is 32 heaps).

**Comment by pshipton** [comment]
ok.

https://ci.eclipse.org/openj9/view/Test/job/Grinder/702 - multiheap:24 x 3
https://ci.eclipse.org/openj9/view/Test/job/Grinder/703 - multiheap:28 x 3
Will start multiheap:30 after some time, too many starting at the same time may cause artifactory download problems.

**Comment by pshipton** [comment]
Trying also the entire extended.system suite
https://ci.eclipse.org/openj9/view/Pipelines/job/Pipeline-Release-Build/178 - multiheap:24
https://ci.eclipse.org/openj9/view/Pipelines/job/Pipeline-Release-Build/179 - multiheap:28

**Comment by pshipton** [comment]
https://ci.eclipse.org/openj9/view/Test/job/Grinder/705 - multiheap:18 x 3
https://ci.eclipse.org/openj9/view/Test/job/Grinder/706 - multiheap:20 x 3

**Comment by zl-wang** [comment]
This is worrisome.  What is the difference between java11 and java8, re the environment?  we also didn't chase further the previous failure with RSS set to unlimited. It is worthwhile to launch a previous failure with everything set to unlimited.

**Comment by pshipton** [comment]
I expect jdk11 just uses more memory. There are many additional classes loaded for modularity on jdk11+, and additional native structures used to manage it.

I don't think setting rss to unlimited made any difference, but I can re-run multiheap:20 with this setting.

**Comment by pshipton** [comment]
https://ci.eclipse.org/openj9/view/Test/job/Grinder/708 - multiheap:20 x 3 unlimited

**Comment by zl-wang** [comment]
i knew setting RSS to unlimited didn't make a difference. So, it must be some other limit. Such that, I suggested setting every limit to unlimited and running one of the known failures.

**Comment by pshipton** [comment]
The only other limit is the stack, which is already at the maximum.
https://github.com/eclipse/openj9/issues/8842#issuecomment-598890366

**Comment by zl-wang** [comment]
ok

**Comment by pshipton** [comment]
https://ci.eclipse.org/openj9/view/Test/job/Grinder/709 - multiheap:24 x 3 unlimited
https://ci.eclipse.org/openj9/view/Test/job/Grinder/710 - multiheap:28 x 3 unlimited
https://ci.eclipse.org/openj9/view/Test/job/Grinder/711 - multiheap:32 x 3 unlimited

**Comment by pshipton** [comment]
Is there a problem to set the default to `multiheap:16`?

**Comment by zl-wang** [comment]
there might be performance implication when the application has many threads.  so that, I asked for:  https://github.com/eclipse/openj9/issues/8842#issuecomment-601719169

**Comment by pshipton** [comment]
Trying all extended.system with multiheap:16
https://ci.eclipse.org/openj9/view/Pipelines/job/Pipeline-Release-Build/181/

**Comment by JasonFengJ9** [comment]
Also occurred at internal builds such as `Test_openjdk14_j9_extended.system_ppc64_aix_Nightly/47`.
```
MT5 01:07:49 >> --> Threads 7500 to 7800 <--
MT3 stderr JVMDUMP039I Processing dump event "systhrow", detail "java/lang/OutOfMemoryError" at 2020/03/22 01:07:52 - please wait.
MT3 stderr JVMDUMP032I JVM requested System dump using '/home/jenkins/workspace/Test_openjdk14_j9_extended.system_ppc64_aix_Nightly/openjdk-tests/TKG/test_output_158484675279/SharedClasses.SCM01.MultiThread_0/20200322-010102-SharedClasses/results/core.20200322.010752.9896074.0001.dmp' in response to an event

MT3 stderr JVMDUMP010I Snap dump written to /home/jenkins/workspace/Test_openjdk14_j9_extended.system_ppc64_aix_Nightly/openjdk-tests/TKG/test_output_158484675279/SharedClasses.SCM01.MultiThread_0/20200322-010102-SharedClasses/results/Snap.20200322.010752.9896074.0004.trc
MT3 stderr JVMDUMP013I Processed dump event "systhrow", detail "java/lang/OutOfMemoryError".
MT3 stderr Exception in thread "main" java.lang.OutOfMemoryError: Failed to create a thread: retVal -1073741830, errno 11
MT3 stderr 	at java.base/java.lang.Thread.startImpl(Native Method)
MT3 stderr 	at java.base/java.lang.Thread.start(Thread.java:1012)
MT3 stderr 	at net.openj9.test.sc.LoaderSlaveMultiThread.run(LoaderSlaveMultiThread.java:99)
MT3 stderr 	at net.openj9.test.sc.LoaderSlaveMultiThread.main(LoaderSlaveMultiThread.java:57)
STF 01:10:38.175 - Found dump at: /home/jenkins/workspace/Test_openjdk14_j9_extended.system_ppc64_aix_Nightly/openjdk-tests/TKG/test_output_158484675279/SharedClasses.SCM01.MultiThread_0/20200322-010102-SharedClasses/results/heapdump.20200322.010802.11468860.0002.phd
STF 01:10:38.177 - **FAILED** Process MT3 ended with exit code (1) and not the expected exit code/s (0)
```

**Comment by shanchao95** [comment]
@zl-wang  multiheap default, 24, 28, 30, 32 performance for 24 and 32 threads [multiheap_performance.xlsx](https://github.com/eclipse/openj9/files/4375629/multiheap_performance.xlsx)

**Comment by zl-wang** [comment]
@shanchao95  thanks for the data. the performance varied quite a bit, especially looking at multiheap default and multiheap:32 (they should be the same or comparable, since default multiheap is 32). do you have the confidence intervals of them?

In addition, I am more interested in multiheap:16, multiheap:18, and multiheap:20 with thread number 16, 24, and 32.  Could you rerun for these?

Notice multiheap:24 (and above) didn't work for this test case beyond java8. so, multiheap:24/28/30/32 kind of meaningless.

**Comment by shanchao95** [comment]
Sure, i'll update when it is done

**Comment by shanchao95** [comment]
@zl-wang performance updated, each test run for 3 times.
[malloc_performance.xlsx](https://github.com/eclipse/openj9/files/4381574/malloc_performance.xlsx)

**Comment by zl-wang** [comment]
Thanks for the data.
My summary for the data:
1) for small payload, the setting can have a significant performance impact.  up to 40% difference between multiheap:16 and multiheap:20 for 32-thread scenario;
2) for large payload, performance impact of specific setting is much less;
3) there are algorithms (eckeyGeneration ones) that the performance variation is so big that no conclusion can be drawn.

Since multiheap:18 appeared safe for both java8 and java11, let's settle for that as a fix.
We need to note somewhere that, for specific performance tuning exercise, multiheap:32 may well be recommended.

**Comment by pshipton** [comment]
@zl-wang the other option is to stick with "multiheap(:32)" and have the failing system testing re-tuned to run a smaller workload. It seems a better solution. I believe the test is running 5 JVMs concurrently.

The concern is that the change may impact a user, in that something that was working may start failing due to OOM. I expect there are still a small number of jdk11 users so far. The benefit of sticking with "multiheap(:32)" is better perf out of the box, and a user may not necessarily go looking for how to change the environment to get better perf. A user getting OOM can either use a beefier machine, or find help to re-tune the MALLOCOPTIONS with the caveat that it may affect perf.

**Comment by zl-wang** [comment]
@pshipton  sure, that is certainly better, on the assumption that system testing can be reset.  I was concerned with that there is failing test that build cannot promote.

let's launch a few sniffing retuned tests with multiheap or multiheap:32 setting.

did you imply that system ran out of physical memory or something by retuning or cutdown?

**Comment by pshipton** [comment]
The failing tests don't block any promotion, they just fail every night in the testing.

> did you imply that system ran out of physical memory or something by retuning or cutdown?

The jdk11+ extended.system suite running now with "multiheap" usually doesn't complete. I see OOMs which aren't just a failure to allocate threads. The testing is getting killed, I'm assuming by the OS.

https://ci.eclipse.org/openj9/job/Test_openjdk11_j9_extended.system_ppc64_aix_Nightly/323
`02:20:43  gmake[4]: *** [autoGen.mk:267: SharedClasses.SCM01.MultiThread_0] Killed`

https://ci.eclipse.org/openj9/job/Test_openjdk14_j9_extended.system_ppc64_aix_Nightly/27
`03:07:52  gmake[3]: *** [/home/u0020236/workspace/Test_openjdk14_j9_extended.system_ppc64_aix_Nightly/openjdk-tests/TKG/../TKG/settings.mk:249: extended.system-sharedClasses] Killed`

**Comment by pshipton** [comment]
@Mesbah-Alam can we re-tune the following tests on jdk11+. With the new MALLOCOPTIONS setting on the system, the machines can't handle the test load. If we aren't already, we can set a -Xmx for each process to reduce the memory usage (such as -Xmx512m). Otherwise we may need to only start 4 JVMs.

SharedClasses.SCM01.MultiThread_0
SharedClasses.SCM23.MultiThread_0

**Comment by zl-wang** [comment]
That mixed failing/killing reasons sounded to me that system is running out of paging space that kernel is selectively killing the processes (usually the one first consuming a lot of paging space). That usually means that system is running on the brink of memory/paging space exhaustion. That made some senses that failing to allocate memory could not be explained previously.

**Comment by pshipton** [comment]
For reference here are some stats from one JVM. The artifact I found only has 1 javacore, although there are 5 core files from different processes. The object heap has only expanded to 155MB, so I'm not sure if a -Xmx512MB will help, but seems we should try that before changing the number of processes.
```
0SECTION       NATIVEMEMINFO subcomponent dump routine
NULL           =================================
0MEMUSER
1MEMUSER       JRE: 3,262,054,024 bytes / 16421 allocations
1MEMUSER       |
2MEMUSER       +--VM: 2,614,356,264 bytes / 10551 allocations
2MEMUSER       |  |
3MEMUSER       |  +--Classes: 87,954,448 bytes / 981 allocations
3MEMUSER       |  |  |
4MEMUSER       |  |  +--Shared Class Cache: 67,108,960 bytes / 2 allocations
3MEMUSER       |  |  |
4MEMUSER       |  |  +--Other: 20,845,488 bytes / 979 allocations
2MEMUSER       |  |
3MEMUSER       |  +--Modules: 2,524,164,336 bytes / 7325 allocations
3MEMUSER       |  |  |
4MEMUSER       |  |  +--Other: 509,328 bytes / 2045 allocations
2MEMUSER       |  |
3MEMUSER       |  +--Memory Manager (GC): 2,193,777,720 bytes / 2766 allocations
3MEMUSER       |  |  |
4MEMUSER       |  |  +--Java Heap: 2,147,483,648 bytes / 1 allocation
3MEMUSER       |  |  |
4MEMUSER       |  |  +--Other: 46,294,072 bytes / 2765 allocations
2MEMUSER       |  |
3MEMUSER       |  +--Threads: 77,796,352 bytes / 1268 allocations
3MEMUSER       |  |  |
4MEMUSER       |  |  +--Java Stack: 1,634,440 bytes / 260 allocations
3MEMUSER       |  |  |
4MEMUSER       |  |  +--Native Stack: 74,579,968 bytes / 261 allocations
3MEMUSER       |  |  |
4MEMUSER       |  |  +--Other: 1,581,944 bytes / 747 allocations
2MEMUSER       |  |
3MEMUSER       |  +--Trace: 2,628,312 bytes / 1037 allocations
2MEMUSER       |  |
3MEMUSER       |  +--JVMTI: 17,776 bytes / 13 allocations
2MEMUSER       |  |
3MEMUSER       |  +--JNI: 64,064 bytes / 121 allocations
2MEMUSER       |  |
3MEMUSER       |  +--Port Library: 249,370,784 bytes / 75 allocations
3MEMUSER       |  |  |
4MEMUSER       |  |  +--Unused <32bit allocation regions: 249,359,280 bytes / 1 allocation
3MEMUSER       |  |  |
4MEMUSER       |  |  +--Other: 11,504 bytes / 74 allocations
2MEMUSER       |  |
3MEMUSER       |  +--Other: 2,237,480 bytes / 2245 allocations
1MEMUSER       |
2MEMUSER       +--JIT: 647,592,160 bytes / 5832 allocations
2MEMUSER       |  |
3MEMUSER       |  +--JIT Code Cache: 268,435,520 bytes / 1 allocation
2MEMUSER       |  |
3MEMUSER       |  +--JIT Data Cache: 16,777,728 bytes / 8 allocations
2MEMUSER       |  |
3MEMUSER       |  +--Other: 362,378,912 bytes / 5823 allocations
1MEMUSER       |
2MEMUSER       +--Class Libraries: 105,600 bytes / 38 allocations
2MEMUSER       |  |
3MEMUSER       |  +--VM Class Libraries: 105,600 bytes / 38 allocations
3MEMUSER       |  |  |
4MEMUSER       |  |  +--sun.misc.Unsafe: 2,656 bytes / 7 allocations
4MEMUSER       |  |  |  |
5MEMUSER       |  |  |  +--Direct Byte Buffers: 528 bytes / 6 allocations
4MEMUSER       |  |  |  |
5MEMUSER       |  |  |  +--Other: 2,128 bytes / 1 allocation
3MEMUSER       |  |  |
4MEMUSER       |  |  +--Other: 102,944 bytes / 31 allocations
NULL           
NULL           ------------------------------------------------------------------------
0SECTION       MEMINFO subcomponent dump routine
NULL           =================================
NULL           
1STHEAPTYPE    Object Memory
NULL           id                 start              end                size               space/region
1STHEAPSPACE   0x0000010010691B70         --                 --                 --         Generational 
1STHEAPREGION  0x0000010010692810 0x0000000080000000 0x0000000087960000 0x0000000007960000 Generational/Tenured Region 
1STHEAPREGION  0x0000010010692230 0x00000000FE560000 0x00000000FFA10000 0x00000000014B0000 Generational/Nursery Region 
1STHEAPREGION  0x0000010010691C50 0x00000000FFA10000 0x0000000100000000 0x00000000005F0000 Generational/Nursery Region 
NULL
1STHEAPTOTAL   Total memory:                   155189248 (0x0000000009400000)
1STHEAPINUSE   Total memory in use:            100966360 (0x0000000006049FD8)
1STHEAPFREE    Total memory free:               54222888 (0x00000000033B6028)
```

**Comment by pshipton** [comment]
This is surprising, I wonder if the usage can be reduced.
`Modules: 2,524,164,336 bytes / 7325 allocations`
344,595 bytes per allocation on average.

**Comment by zl-wang** [comment]
module is the memory hogger here, in addition to invisible thread system stacks maybe.

**Comment by zl-wang** [comment]
how to understand the discrepancy:  modules + GC look like more than 4GB already, but JRE is 3.2GB?

**Comment by pshipton** [comment]
> how to understand the discrepancy: modules + GC look like more than 4GB already, but JRE is 3.2GB?

yes, it doesn't make sense, perhaps there is a bug in how the accounting for modules is set up. 

The test is creating and using 9000 jar files. Mesbah tried 5000 and it still failed, but 500 worked. He'll try other values between 500 and 5000 and we will adjust the test.

**Comment by zl-wang** [comment]
it still means that multiheap setting increases RSS, at least memory consumption (eventually over the paging space).  I am wondering, for this test case, how much the increase is.

**Comment by pshipton** [comment]
> I am wondering, for this test case, how much the increase is.

How do we determine RSS on AIX? It's a bit tricky to get RSS from a running process at a consistent time. I can ask Mesbah to modify the test to pause at the end while RSS is captured.

**Comment by zl-wang** [comment]
> How do we determine RSS on AIX? It's a bit tricky to get RSS from a running process at a consistent time.

ps -lfp <pid>
and, looking at the SZ column/field.  That is RSS in kbyte.

**Comment by pshipton** [comment]
We used the 0.18.1 release build. I'll fill in the additional data as it becomes available.

No MALLOCOPTIONS
```
14:01:10  MT         F S      UID      PID     PPID   C PRI NI ADDR    SZ    WCHAN    STIME    TTY  TIME CMD
14:01:10  MT    740001 A  jenkins  8781984 12648466  58  60 20 865306590 130108        * 13:59:44      -  0:58 
```

**Comment by zl-wang** [comment]
unfortunately, that is a big increase (even for 18 heap).  it might be specific to this test case allocating so many threads.  can we have a comparison on Liberty for example? 

sorry, @pshipton    i misremembered the ps command flag, although SZ and RSS are close mostly.  the real flag for RSS is v:

ps  v  [pid]

**Comment by zl-wang** [comment]
with v flag, there is an explicit RSS field

**Comment by pshipton** [comment]
I added another run using an option I spotted, "considersize", which looks much better but not sure how it affects perf. We'll change the ps option to 'v' and deliver the change to the test. I can do more runs of this test next week.

For Liberty, pls ask the perf team.

Pls ask the perf team for runs of Liberty.

**Comment by zl-wang** [comment]
@shanchao95 could you run a performance comparison between "multiheap"  and "multiheap,considersize" settings?

**Comment by zl-wang** [comment]
@vijaysun-omr      @piyush  could you do a footprint size comparison on AIX running Liberty/DT7 with these three environment variable settings below? 
1) no setting
2) export MALLOCOPTIONS=multiheap
3) export MALLOCOPTIONS=multiheap,considersize

**Comment by pshipton** [comment]
@piyush note the latest OpenJ9 JVMs already set "MALLOCOPTIONS=multiheap". It not sure if `export MALLOCOPTIONS=` overrides that. I suggest using the 0.18.1 release JVM to run the tests.

**Comment by pshipton** [comment]
Not exactly apples to apples, but the result is consistent with the previous runs, and difference is so big I don't think it matters.

jdk11 `MALLOCOPTIONS=`
```
10:50:08  MT1       PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
10:50:08  MT1  16056544      - A     4:02 1260 167540 167668 65536    75   128  4.8  3.0 /home/
10:50:08  MT3 stdout of command: ps v 15401100
10:50:08  MT3       PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
10:50:08  MT3  15401100      - A     4:19 1037 166964 167092 65536    75   128  5.1  3.0 /home/
10:50:10  MT1 stderr of command: ps v 16056544
10:50:10  MT3 stderr of command: ps v 15401100
10:50:11  MT5       PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
10:50:11  MT5   8323326      - A     4:10 6984 159348 159476 65536    75   128  4.9  3.0 /home/
10:50:11  MT5 stderr of command: ps v 8323326
10:50:13  MT2 stdout of command: ps v 12124166
10:50:13  MT2       PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
10:50:13  MT2  12124166      - A     4:17  939 169396 169524 65536    75   128  5.0  3.0 /home/
10:50:13  MT2 stderr of command: ps v 12124166
10:50:14  MT4 stdout of command: ps v 16318662
10:50:14  MT4       PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
10:50:14  MT4  16318662      - A     3:48  875 149428 149556 65536    75   128  4.4  2.0 /home/
```

jdk14 `MALLOCOPTIONS=multiheap`
```
[2020-03-29T17:37:25.876Z] MT3      PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
[2020-03-29T17:37:25.876Z] MT3  2818280      - A     4:06 1150 758264 734456 65536    80   128  4.8  9.0 /home/u
[2020-03-29T17:37:25.876Z] MT3 stderr of command: ps v 2818280
[2020-03-29T17:37:25.876Z] MT1 stdout of command: ps v 2556078
[2020-03-29T17:37:25.876Z] MT1      PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
[2020-03-29T17:37:25.876Z] MT1  2556078      - A     4:12 1167 839672 808056 65536    80   128  4.9 10.0 /home/u
[2020-03-29T17:37:25.876Z] MT1 stderr of command: ps v 2556078
[2020-03-29T17:37:25.876Z] MT4 stdout of command: ps v 6947044
[2020-03-29T17:37:25.876Z] MT4      PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
[2020-03-29T17:37:25.876Z] MT4  6947044      - A     4:09 1471 813688 779320 65536    80   128  4.8  9.0 /home/u
[2020-03-29T17:37:28.180Z] MT5 stdout of command: ps v 5439564
[2020-03-29T17:37:28.180Z] MT5      PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
[2020-03-29T17:37:28.180Z] MT5  5439564      - A     4:17 1496 916536 889016 65536    80   128  5.0 11.0 /home/u
[2020-03-29T17:37:28.180Z] MT5 stderr of command: ps v 5439564
[2020-03-29T17:37:28.180Z] MT4 stderr of command: ps v 6947044
[2020-03-29T17:37:32.163Z] MT2 stdout of command: ps v 5636118
[2020-03-29T17:37:32.872Z] MT2      PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
[2020-03-29T17:37:32.872Z] MT2  5636118      - A     4:11 4253 855096 823992 65536    80   128  4.8 10.0 /home/u
```

**Comment by pshipton** [comment]
jdk11 `MALLOCOPTIONS=multiheap,considersize`
```
11:22:03  MT3 stdout of command: ps v 6226366
11:22:03  MT3      PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
11:22:03  MT3  6226366      - A     4:10 4607 157172 157300 32768    75   128  4.9  2.0 /home/u
11:22:03  MT3 stderr of command: ps v 6226366
11:22:03  MT1 stdout of command: ps v 7799084
11:22:03  MT5 stdout of command: ps v 7405844
11:22:03  MT5      PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
11:22:03  MT5  7405844      - A     4:17  772 166196 166324 32768    75   128  5.1  2.0 /home/u
11:22:03  MT1      PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
11:22:03  MT1  7799084      - A     4:04 2714 158452 158580 32768    75   128  4.8  2.0 /home/u
11:22:03  MT1 stderr of command: ps v 7799084
11:22:03  MT5 stderr of command: ps v 7405844
11:22:06  MT2 stdout of command: ps v 4718840
11:22:06  MT2      PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
11:22:06  MT2  4718840      - A     4:15  719 164980 165108 32768    75   128  4.9  2.0 /home/u
11:22:06  MT2 stderr of command: ps v 4718840
11:22:09  MT4 stdout of command: ps v 8519920
11:22:09  MT4      PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
11:22:09  MT4  8519920      - A     4:04  723 157684 157812 32768    75   128  4.7  2.0 /home/u
11:22:09  MT4 stderr of command: ps v 8519920
```

**Comment by zl-wang** [comment]
nice.  now, @shanchao95 performance comparison is the critical factor for fix decision.

**Comment by pshipton** [comment]
jdk 11 `MALLOCOPTIONS=multiheap:18`
```
12:01:47  MT5 stdout of command: ps v 11927744
12:01:47  MT5       PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
12:01:47  MT5  11927744      - A     4:10 1501 685620 685748 65536    75   128  5.0  8.0 /home/
12:01:47  MT5 stderr of command: ps v 11927744
12:01:47  MT1 stdout of command: ps v 16711774
12:01:47  MT1       PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
12:01:47  MT1  16711774      - A     4:05 6352 705204 705332 65536    75   128  4.8  9.0 /home/
12:01:47  MT1 stderr of command: ps v 16711774
12:01:51  MT2 stdout of command: ps v 16253166
12:01:51  MT2       PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
12:01:51  MT2  16253166      - A     4:05  840 677428 677556 65536    75   128  4.8  8.0 /home/
12:01:51  MT2 stderr of command: ps v 16253166
12:01:54  MT3 stdout of command: ps v 16056322
12:01:56  MT3       PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
12:01:56  MT3  16056322      - A     4:15  773 590900 591028 65536    75   128  4.9  7.0 /home/
12:01:56  MT3 stderr of command: ps v 16056322
12:01:57  MT4 stdout of command: ps v 14090452
12:01:57  MT4       PID    TTY STAT  TIME PGIN  SIZE   RSS   LIM  TSIZ   TRS %CPU %MEM COMMAND
12:01:57  MT4  14090452      - A     4:20 1134 588340 588468 65536    75   128  5.0  7.0 /home/
12:01:57  MT4 stderr of command: ps v 14090452
```

**Comment by piyush286** [comment]
@zl-wang Sorry I didn't see this issue since my external handle is wrong there. 

@sophiaxu0424 Can we please get the data requested in https://github.com/eclipse/openj9/issues/8842#issuecomment-605317393? As Peter recommend in https://github.com/eclipse/openj9/issues/8842#issuecomment-605323016, we should use the 0.18.1 release build. Thanks!

**Comment by shanchao95** [comment]
@zl-wang Sure, here is performance for `MALLOCOPTIONS=multiheap,considersize`, `MALLOCOPTIONS=multiheap:18` and `MALLOCOPTIONS=`.
[performance.xlsx](https://github.com/eclipse/openj9/files/4409533/performance.xlsx)

**Comment by zl-wang** [comment]
@shanchao95  thanks for the data.  but I am confused:  previously between multiheap setting and no-setting, for 16-thread tests on small 512 payload, the performance improvement was from a few times to 100s times. however, this batch of data didn't show that at all.

I am wondering if your driver is after the code merge such that your no-setting actually means the current setting in the merge.

**Comment by shanchao95** [comment]
@zl-wang I found a different version of jdk was used for this performance test. Previously i used jvm**0317 from espresso. A new run with jdk0317 is running. It will be updated once it is done.

**Comment by zl-wang** [comment]
@shanchao95  thanks a lot for the data.  summary:
1) multiheap,considersize is 50% to 70% of performance of multiheap which is 10x to 100x performance of no-setting, for multi-threaded cases;
2) multiheap,consider memory consumption/RSS is in the same ballpark as no-setting. However, multiheap RSS  have two cases:
    2a) intensive multi-thread malloc/free:  it is more than 5x of no-setting RSS (e.g.  this PR test case)
    2b) sporadic malloc/free:  no impact ... in the same ballpark as no-setting (e.g. DT7 test)

what is the converged fix?  that is a good question. I inclined to multiheap,considersize for moderate RSS plus most of the performance benefit.  Note somewhere: multiheap,bucket setting may give another 100% to 40% performance boost.

**Comment by pshipton** [comment]
@zl-wang we need to finalize the setting to be used for the 0.20.0 release this week. For the time being, I'm going to change to the head stream to "multiheap,considersize" to avoid failing the tests every night. 

It sounds like you are considering a perf/footprint comparison between `multiheap,considersize` and `multiheap,bucket,considersize`. Can you make an explicit request.

**Comment by vijaysun-omr** [comment]
I am fine with the proposed `considersize` setting but should we get the JCE performance team to also measure with this setting on AIX to get the performance picture on the multithreaded scenarios they were running on ?

**Comment by zl-wang** [comment]
@pshipton  sorry for my ambiguity.  additional bucket option only had marginal performance advantage.  i was referring to the extreme case of tuning in performance guide or something. i was proposing multiheap,considersize as the fix.

@vijaysun-omr   JCE performance data I summarized here:  https://github.com/eclipse/openj9/issues/8842#issuecomment-609841411

**Comment by sophiaxu0424** [comment]
Starup/footpirnt runs have finished, please see data here: https://github.ibm.com/runtimes/javanext/issues/176#issuecomment-19050257. 

Setting-2 keeps failing the jobs and might need some help to figure it out @zl-wang @piyush286  FYI.

**Comment by pshipton** [comment]
The change to use "multiheap,considersize" is merged to the head stream, but not yet to the 0.20.0 branches

**Comment by pshipton** [comment]
The changes are merged to the 0.20.0 branches. Closing this issue, a new issue can be opened for any further improvements.

**Comment by sophiaxu0424** [comment]
SPECjbb2015 runs also finished, please refer to :  javanext/issues/176 for details.

Setting-1 | max_jOPS | critical_jOPS
-- | -- | --
mean | 19036.5 | 10387.5
CI | 2.72% | 4.41%
  |   |  
  |   |  
**Setting**-2 | max_jOPS | critical_jOPS
mean | 38467.75 | 15120.25
CI | 0.91% | 3.14%
  |   |  
**setting**-2/**setting**-1 | **202.07**% | **145.56**%
  |   |  
**Setting**-3 | max_jOPS | critical_jOPS
mean | 39093 | 15683
CI | 0.00% | 2.42%
  |   |  
**setting**-3/**setting**-1 | **205.36**% | **150.98**%
